code/game/gamemodes/malfunction/Malf_Modules.dm

```diff
@@ -1,51 +1,274 @@
 	var/cost = 5
-	var/one_time = 0
+	var/one_purchase = 0 //If this module can only be purchased once. This always applies to upgrades, even if the variable is set to false.
 
-	var/power_type
+	var/power_type = /datum/action/innate/ai //If the module gives an active ability, use this. Mutually exclusive with upgrade.
+	var/upgrade //If the module gives a passive upgrade, use this. Mutually exclusive with power_type.
+	var/unlock_text = "<span class='notice'>Hello World!</span>" //Text shown when an ability is unlocked
+	var/unlock_sound //Sound played when an ability is unlocked
 
-/datum/AI_Module/large
-	uses = 1
+/datum/AI_Module/proc/upgrade(mob/living/silicon/AI/AI) //Apply upgrades!
+	return
 
-/datum/AI_Module/small
-	uses = 5
+/datum/AI_Module/large //Big, powerful stuff that can only be used once.
+/datum/AI_Module/small //Weak, usually localized stuff with multiple uses.
 
+
+//Doomsday Device: Starts the self-destruct timer. It can only be stopped by killing the AI completely.
 /datum/AI_Module/large/nuke_station
 	module_name = "Doomsday Device"
 	mod_pick_name = "nukestation"
 	description = "Activate a weapon that will disintegrate all organic life on the station after a 450 second delay. Can only be used while on the station, will fail if your core is moved off station or destroyed."
 	cost = 130
-	one_time = 1
-
-	power_type = /mob/living/silicon/ai/proc/nuke_station
-
-/mob/living/silicon/ai/proc/nuke_station()
-	set category = "Malfunction"
-	set name = "Doomsday Device"
-
-	var/turf/T = get_turf(src)
-
+	one_purchase = TRUE
+	power_type = /datum/action/innate/ai/nuke_station
+	unlock_text = "<span class='notice'>You slowly, carefully, establish a connection with the on-station self-destruct. You can now activate it at any time.</span>"
+
+/datum/action/innate/ai/nuke_station
+	name = "Doomsday Device"
+	desc = "Activates the doomsday device. This is not reversible."
+	button_icon_state = "doomsday_device"
+	auto_use_uses = FALSE
+
+/datum/action/innate/ai/nuke_station/Activate()
+	var/turf/T = get_turf(owner)
 	if(!istype(T) || T.z != ZLEVEL_STATION)
-		to_chat(src, "<span class='warning'>You cannot activate the doomsday device while off-station!</span>")
+		to_chat(owner, "<span class='warning'>You cannot activate the doomsday device while off-station!</span>")
 		return
-
-	to_chat(src, "<span class='notice'>Doomsday device armed.</span>")
-	priority_announce("Hostile runtimes detected in all station systems, please deactivate your AI to prevent possible damage to its morality core.", "Anomaly Alert", 'sound/ai/aimalf.ogg')
+	if(alert(owner, "Send arming signal? (true = arm, false = cancel)", "purge_all_life()", "confirm = TRUE;", "confirm = FALSE;") != "confirm = TRUE;")
+		return
+	active = TRUE
+	INVOKE_ASYNC(src, .proc/set_us_up_the_bomb, owner)
+
+/datum/action/innate/ai/nuke_station/proc/set_us_up_the_bomb(mob/living/owner)
+	to_chat(owner, "<span class='small boldannounce'>run -o -a 'selfdestruct'</span>")
+	sleep(5)
```

Comment:
That seems like it'd make the proc really inefficient, so I'll see if I can make it better with a forloop and iterating through a list of functions.